Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+    where 
+        lA                  = length polyA 
+        lB                  = length polyB
+        leadingZerosToA 
+            | lA < lB       = lB - lA
+            | otherwise     = 0
+        leadingZerosToB 
+            | lB < lA       = lA - lB
+            | otherwise     = 0
+
+-- Exercise 5 -----------------------------------------
+
+-- c is the coeff being multiplied from the first polynomial
+-- e is the exponentioal term being used for multiplication from first polynomial
+
+-- timesHelper (P [1, 4, 2]) 1 0 == P [1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 1 == P [0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 2 == P [0, 0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 2 2 == P [0, 0, 2, 8, 4]
+
+-- 1. add leading zeros list of size == e 
+-- and multiply rest with c
+timesHelper ::  (Num a)     => Poly a -> a -> Int -> Poly a
+timesHelper (P p) c e       = P $ replicate e 0 ++ map (*c) p
+
+-- getAllTimes 0 (P [1, 1, 2]) (P [1, 4, 2]) == [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]]
+getAllTimes ::  (Num a)     => Int -> Poly a -> Poly a -> [Poly a]
+getAllTimes _ (P []) _      = []
+getAllTimes n (P [p]) q     = [timesHelper q p n]
+getAllTimes n (P (p:ps)) q  = timesHelper q p n :  getAllTimes (n+1) (P ps) q
+
+-- addAllTimes [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]] == P [1, 5, 8, 10, 4]
+-- This is how it works with the help of [timesHelper] function
+--      [1, 4, 2, 0, 0]
+--   +  [0, 1, 4, 2, 0]
+--   +  [0, 0, 2, 8, 4]
+--      ---------------
+--      [1, 5, 8, 10, 4]
+--      ---------------
+addAllTimes :: (Num a)      => [Poly a] -> Poly a
+addAllTimes []              = P []
+addAllTimes [p1]            = p1
+addAllTimes [p1, p2]        = plus p1 p2
+addAllTimes (p1:p2:ps)      = plus p1 p2 + addAllTimes ps
+
+times :: (Num a)            => Poly a -> Poly a -> Poly a
+times p q                   = addAllTimes $ getAllTimes 0 p q
+
+-- Exercise 6 -----------------------------------------
+
+instance Num a             => Num (Poly a) where
+    (+)                    = plus
+    (*)                    = times
+    negate                 = (* P [-1])
+    fromInteger i          = P [fromInteger i]
+
+    -- No meaningful definitions exist
+    abs    = undefined
+    signum = undefined
+
+-- Exercise 7 -----------------------------------------
+
+applyP :: Num a            => Poly a -> a -> a
+applyP (P p) valX          = sum $ zipWith (\a b -> a * (valX ^ b)) p intList
+    where 
+        intList :: [Int]
+        intList            = [0..]
+
+-- Exercise 8 -----------------------------------------
+
+-- I wasnt able to understand what is nderiv here so this is the only exc left in this weeks assignment 
+class Num a => Differentiable a where
+    deriv  :: a -> a
+    nderiv :: Int -> a -> a
+    nderiv = undefined
+
+-- Exercise 9 -----------------------------------------
+
+getDerivConst :: (Num a, Eq a) => a -> a -> a
+getDerivConst c e
+    | e == 0               = 0
```

Comment:
If 'e == 0' then 'c * e == 0', so no need for branching here

Reply:
right, i don't even need this function i guess. can just do like this

**Before**
```
deriv (P (_:ps))        = P $ zipWith getDerivConst ps intList
            where 
                intList     = [1..]
```

**After**
```
deriv (P (_:ps))        = P $ zipWith (*) ps intList
            where 
                intList     = [1..]
```

Reply:
Right